errors:

# -------------------------- #
#       Database Errors      #
# -------------------------- #

  - message: "Could not list databases/tables for connection."
    id: "list-databases-tables"
    applies-to: "databases"
    ip-addresses: false
    meaning: "Stitch is unable to access the databases or tables in your database."
    fix-it: &verify-permissions |
      Verify that the Stitch database user has the necessary permissions for the type of database you’re connecting.

      You can find the permissions in the [Creating a Stitch Database User section of the setup guide for your database](/docs/integrations/databases).


  - message: "[table_name] unsupported"
    id: "unsupported-table"
    applies-to: "databases"
    ip-addresses: false
    meaning: "Stitch is unable to access the table or the table is in an unsupported format."
    fix-it: *verify-permissions


  - message: "Prematurely reached end of file/stream"
    id: "premature-file"
    applies-to: "databases"
    ip-addresses: false
    meaning: |
      **Applicable only to MongoDB integrations,** this error usually means that SSL has been incorrectly configured.

      Connecting a database integration to Stitch via SSL has two parts: configuration on the database's server and in the Stitch app. For the connection to be successful, the settings in both Stitch and on the database server must align.

      For example: a MongoDB server doesn't support SSL connections but the SSL option is checked in Stitch. This will result in a connection error.
    fix-it: "First, verify if the MongoDB server is configured to support SSL connections. Then:"
    fix-it-list:
      - item: "**If SSL connections aren't supported,** make sure the **Connect using SSL** box in Stitch is **unchecked** and try saving the integration again."
      - item: "**If SSL connections are required,** make sure the **Connect using SSL** box in Stitch is **checked** and try saving the integration again."


  - message: |
      ...Command failed with error 18: 'Authentication failed.' on server [[host]:[port]...
    full-message: |
      Timed out after XXXX ms while waiting for a server that matches ReadPreferenceServerSelector{readPreference=primary}.

      Client view of cluster state is {type=UNKNOWN, servers=[{address=[host]:[port], type=UNKNOWN, state=CONNECTING, exception={com.mongodb.MongoSecurityException:

      Exception authenticating MongoCredential{mechanism=null, userName='[user_name]', source='[database_name]', password=, mechanismProperties={}}}, caused by {com.mongodb.MongoCommandException

      Command failed with error 18: 'Authentication failed.' on server [[host]:[port].

      The full response is { "ok" : 0.0, "errmsg" : "Authentication failed.", "code" : 18, "codeName" : "AuthenticationFailed" }}}]
    id: "mongo-authentication-database"
    applies-to: "databases"
    ip-addresses: false
    meaning: |
      **Applicable only to MongoDB integrations**, this error usually means that the database entered into the **Database** field of the {{ app.page-names.int-settings }} page is not the same database that's used for user authentication.

      Stitch requires a database user and a MongoDB server with `auth` enabled to create MongoDB connections. When authenticating a user in this way, [MongoDB requires that the user be validated against their authentication database](https://docs.mongodb.com/manual/core/security-users/#authenticate-a-user).

      Currently, Stitch's MongoDB integration currently assumes the database entered in the Integration Settings page is the authentication database. Due to MongoDB's authentication process and Stitch's current functionality, the database entered into Stitch must be the user's authentication database.

      **Note**: This database is only used to create the connection. Stitch will be able to detect any other databases the Stitch user has access to.
    fix-it-list:
      - item: |
          Ensure that the Stitch database user can perform all the required actions on the authentication database, in addition to any other databases that contain data you want to replicate. [See the Create a Mongo Database user section]({{ site.baseurl }}/integrations/databases/mongodb#create-database-user) of the Mongo setup guide for the full list of requirements.

          **Note**: If this is an Atlas-based MongoDB instance, the authentication database will be the `admin` database.

      - item: |
          Once the Stitch user's permissions have been verified, re-enter the name of the database into the {{ app.page-names.int-settings }} page:

            - **For Atlas-based MongoDB instances**, this must be `admin`.
            - **For other MongoDB instances**, enter the name of the authentication database.

      - item: "When finished, save the integration."

# -------------------------- #
#     Destination Errors     #
# -------------------------- #

  - message: "An error occurred (404) when calling the HeadObject operation: Not Found"
    id: "missing-challenge-file"
    applies-to: "destinations"
    ip-addresses: false
    meaning: |
      **Applicable to Amazon S3 destinations only.** This error occurs when the Stitch [challenge file]({{ link.destinations.setup.amazon-s3 | append: "#verify-bucket-access" }}) isn't found in the specified S3 bucket.

      The challenge file is used to test Stitch's connection to the bucket. Stitch will surface this error if it cannot find the challenge file, whether because the file was removed or never created.
    fix-it: |
      - **If you still have access to the challenge file name**, follow [these instructions]({{ link.destinations.setup.amazon-s3 | prepend: site.baseurl | append: "#verify-bucket-access" }}) to create the file in your S3 bucket.
      - **If you no longer have access to the challenge file name**, you'll need to create a new S3 destination [using these instructions]({{ link.destinations.overviews.switch-destination | prepend: site.baseurl }}).

         **Note**: Creating a new destination doesn't mean you'll need to re-replicate all of your data. During the destination change process, you can choose to pick up replication where it left off or to queue a historical re-replication.


# -------------------------- #
#     Applicable to Both     #
# -------------------------- #

  - message: "Connection refused. Check that the hostname and port are correct and that the postmaster is accepting TCP/IP connections."
    id: "connection-refused"
    applies-to: "databases,destinations"
    ip-addresses: true
    meaning: "Stitch encountered an error when attempting to 'talk' to your [destination/database]."
    fix-it-list:
      - item: &hostname |
          Verify that the **hostname** entered in the [Destination/Integration] Settings page is correct.

          **For [types]**: check that the port has been removed from the end of the hostname string.
      - item: &port-setting "Verify that the **port setting** entered in the [Destination/Integration] Settings page is correct."
      - item: &ip-addresses |
          Verify that Stitch’s IP addresses (see below) are whitelisted.

          **For [types]**: Check that the Security Group and VPC Subnet associated with the database allows access from all Stitch's IP addresses.



  - message: "We encountered an error while attempting to check the type or version for this [[destination/database]]. Please select the correct [destination/database] type and try again."
    id: "database-type"
    applies-to: "databases,destinations"
    ip-addresses: false
    meaning: |
      The [destination/database] type doesn’t match the [destination/database] you selected in Stitch. 

      For example: you select [wrong-type] in Stitch, but we detect that it’s actually a PostgreSQL [destination/database].
    fix-it: "Verify that you’ve selected the correct [destination/database] type. If the incorrect [destination/database] type was chosen, select the correct [destination/database] type and try completing the setup again."



  - message: "An I/O error occurred while sending to the backend."
    id: "io-error"
    applies-to: "databases,destinations"
    ip-addresses: false
    meaning: "Something went wrong during the communication with the database."
    fix-it: |
      This is a transient error and should clear itself during the next replication job. If this error persists, refer to [these troubleshooting tips]({{ link.troubleshooting.aws-io-errors | prepend: site.baseurl }}) to diagnose and treat the issue.



  - message: "FATAL: Password authentication failed for user [user_name]"
    id: "password-auth"
    applies-to: "databases,destinations"
    ip-addresses: false
    meaning: "The credentials that Stitch has for your [destination/database] are incorrect."
    fix-it: "Verify that the correct password for the user Stitch is using to connect to the [destination/database] is entered in the [Destination/Integration] Settings page of Stitch."



  - message: "Connection timed out."
    id: "time-out"
    applies-to: "databases,destinations"
    ip-addresses: true
    meaning: "The connection timed out before Stitch could successfully establish a connection."
    fix-it: "This error may be transient and resolve on its own, but if it persists, try the following before contacting support:"
    fix-it-list:
      - item: *port-setting
      - item: *ip-addresses



  - message: "We encountered an internal error while checking the connection."
    id: "internal-error"
    applies-to: "databases,destinations"
    ip-addresses: false
    meaning: "Stitch encountered an internal issue."
    fix-it: "This error is usually transient and will resolve on its own, but if it persists, please check our [Status Page]({{ site.status }}) for reported service outages."